sigstore/_internal/trustroot.py

```diff
@@ -102,72 +106,116 @@ class KeyringSignatureError(KeyringError):
     """
 
 
-class Keyring:
+@dataclass(init=False)
+class Key:
     """
-    Represents a set of CT signing keys, each of which is a potentially
-    valid signer for a Signed Certificate Timestamp (SCT).
-
-    This structure exists to facilitate key rotation in a CT log.
+    Represents a key in a `Keyring`.
     """
 
-    def __init__(self, keys: List[bytes] = []):
+    hash_algorithm: hashes.HashAlgorithm
+    key: PublicKey
+    key_id: KeyID
+
+    _RSA_SHA_256_DETAILS: ClassVar[set[_PublicKeyDetails]] = {
+        _PublicKeyDetails.PKCS1_RSA_PKCS1V5,
+        _PublicKeyDetails.PKIX_RSA_PKCS1V15_2048_SHA256,
+        _PublicKeyDetails.PKIX_RSA_PKCS1V15_3072_SHA256,
+        _PublicKeyDetails.PKIX_RSA_PKCS1V15_4096_SHA256,
+    }
+
+    _EC_DETAILS_TO_HASH: ClassVar[dict[_PublicKeyDetails, hashes.HashAlgorithm]] = {
+        _PublicKeyDetails.PKIX_ECDSA_P256_SHA_256: hashes.SHA256(),
+        _PublicKeyDetails.PKIX_ECDSA_P384_SHA_384: hashes.SHA384(),
+        _PublicKeyDetails.PKIX_ECDSA_P521_SHA_512: hashes.SHA512(),
+    }
+
+    def __init__(self, public_key: _PublicKey) -> None:
         """
-        Create a new `Keyring`, with `keys` as the initial set of signing
-        keys. These `keys` can be in either DER format or PEM encoded.
+        Construct a key from the given Sigstore PublicKey message.
         """
-        self._keyring = {}
-        for key_bytes in keys:
-            key = None
 
-            try:
-                key = load_pem_public_key(key_bytes)
-            except UnexpectedKeyFormatError as e:
-                raise e
-            except InvalidKeyError:
-                key = load_der_public_key(key_bytes)
+        hash_algorithm: hashes.HashAlgorithm
+        if public_key.key_details in self._RSA_SHA_256_DETAILS:
+            hash_algorithm = hashes.SHA256()
+            key = load_der_public_key(public_key.raw_bytes, types=(rsa.RSAPublicKey,))
+        elif public_key.key_details in self._EC_DETAILS_TO_HASH:
+            hash_algorithm = self._EC_DETAILS_TO_HASH[public_key.key_details]
+            key = load_der_public_key(
+                public_key.raw_bytes, types=(ec.EllipticCurvePublicKey,)
+            )
+        else:
+            raise InvalidKeyError(f"unsupported key type: {public_key.key_details}")
```

Comment:
NB: This brings us to parity with the previous behavior, i.e. allowing ECDSA and RSA. But it gives us the flexility to allow additional keytypes in the future 🙂

Reply:
+1 to a nice error message